Ipset pf support

wcawijngaards opened the pull request:
This cleans up the code, after discussion with my colleague, we decided to rename the module calls differently. This makes that change. It also performs a merge with the recent code base.

Fixes #144 .

Commit:
allow privileged initialisation of modules

Commit:
Port ipset to BSD pf tables

Commit:
Don't try to run daemon_privileged on reload.

Commit:
Merge remote-tracking branch 'upstream/master'

wcawijngaards committed:
Merge branch 'master' of https://github.com/madroach/unbound into ipset-pf-support

wcawijngaards committed:
ipset-pf-support, fix compilation, close of pf socket.

wcawijngaards committed:
Call module init init again, and new function startup and destartup.
NULL can be used if the function is not used. Open shared ports during
reload. Deinit is called during reload.

wcawijngaards committed:
ipset-pf-support, fix to remove unused include, free at end, adjust
qname for comparison.

wcawijngaards committed:
ipset-pf-support, fix to log error on failure to open pf.

wcawijngaards committed:
- ipset-pf-support, fix to skip unit test if no pf dev.

wcawijngaards committed:
ipset-pf-support, simplification of code.

gthess commented on ipset/ipset.c:
```diff
@@ -178,23 +272,24 @@ ipset_check_zones_for_rrset(struct module_env *env, struct ipset_env *ie,
+#ifdef HAVE_NET_PFVAR_H
+#else
+	if (!ie->dev) {
+		// retry to create mnl socket
```
```suggestion
		/* retry to create mnl socket */
```

gthess commented on ipset/ipset.c:
```diff
@@ -41,24 +51,95 @@ static int error_response(struct module_qstate* qstate, int id, int rcode) {
 		return NULL;
 	}
-	return mnl;
+	return dev;
```
```suggestion
	return (void *)dev;
```

gthess reviewed (commented):
Some nits in the suggestions.
But some additional points were raised during review:
- In the module function block I would prefer the _startup functions to be before the _init ones to match the perceived execution order (nit).
- Now that we have `modstack_(de)startup` I would like `modstack_(de)setup` to be renamed to `modstack_(de)init`, and `modstack_init` to be renamed to `modstack_something`? But then I think that the better names would be `modstack_call_(de)startup` and `modstack_call_(de)init`. Then `modstack_init` can just stay the same. That way we have distinct naming when modstack methods are for the modstack struct itself and when they are calling module functions.
- I think for the reordering of the modules during reload, we can block only if any module has a startup function, otherwise we don't care, right?

wcawijngaards committed:
Apply suggestions from code review

Co-authored-by: Yorgos Thessalonikefs <yorgos@nlnetlabs.nl>

wcawijngaards committed:
ipset-pf-support, move startup and destartup to the front of the module
func block functions, modstack call deinit function names, and detect
module change when no startup functions are needed.

wcawijngaards commented:
Added a commit that fixes those review issues. The naming like `modstack_call_deinit` is nice. I have also added `modstack_free`. That is also used when it detects a change in module list that does not involve modules that need `startup`. The modstack deinit was called too often, during the first start as well, the `daemon->mods_inited` variable tracks that, so that deinit is called during reloads, but not when the server starts for the first time.

gthess reviewed (approved):
This looks good to me now, thanks!
